feat(plugins/opentelemetry): Replace PeriodicReader with direct gRPC export (#400)
alumet-dev/alumet: plugins/opentelemetry/README.md

This crate is a library that defines the OpenTelemetry plugin.

> Implements a push-based exporter (via gRPC) which can be connected to an OpenTelemetry Collector (via a receiver), processed in any way, and then exported to an observability backend like Jaeger, Prometheus, Thanos, OpenSearch, ElasticSearch, etc.

Reviewer: ```suggestion
Implements a push-based exporter (via OTLP/gRPC) which can be connected to an OpenTelemetry Collector (via a receiver), processed in any way, and then exported to an observability backend like Jaeger, Prometheus, Thanos, OpenSearch, ElasticSearch, etc.
```

Each call to `write()` triggers an immediate gRPC export, allowing metrics to be exported at any frequency.

## Requirements

> - A reachable OpenTelemetry Collector with a gRPC receiver enabled (default port: `4317`).

Reviewer: ```suggestion
- A reachable OpenTelemetry Collector with an OTLP/gRPC receiver enabled.
```

```toml
[plugins.opentelemetry]
# URL of the OTLP gRPC collector
collector_host = "http://localhost:4317"
# Optional prefix and suffix applied to every exported metric name
prefix = ""
suffix = "_alumet"
# Use the display name of the units instead of their unique name, as specified by the UCUM.
# See https://ucum.org/ucum for a list of units and their symbols.
# Example: "W" (display) vs "watt" (unique)
use_unit_display_name = true
# Forward measurement attributes as OpenTelemetry data-point attributes
add_attributes_to_labels = true
```

> 

Reviewer: It would be nice to add a section in "More Information" that explain how frequently the requests will be sent to the otel collector, explaining that it depends on the flush_interval of sources  (but explaining that it's decoupled from the poll_interval concept etc)

```toml
[plugins.opentelemetry]
# URL of the OTLP gRPC collector
collector_host = "http://localhost:4317"
# Optional prefix and suffix applied to every exported metric name
prefix = ""
suffix = "_alumet"
# Use the display name of the units instead of their unique name, as specified by the UCUM.
# See https://ucum.org/ucum for a list of units and their symbols.
# Example: "W" (display) vs "watt" (unique)
use_unit_display_name = true
# Forward measurement attributes as OpenTelemetry data-point attributes
add_attributes_to_labels = true
```

## More information

Check more at the [user-book website](https://alumet-dev.github.io/user-book/plugins/output/opentelemetry.html).

> > **Note**: If a point with an empty value is received, this plugin will set it's value to `"empty"` before forwarding it to OpenTelemetry.

Reviewer: ```suggestion
> **Note**: If a point with an empty attribute value is received, this plugin will set it's value to `"empty"` before forwarding it to OpenTelemetry.
```

The resulting traffic seen by the OTEL Collector looks like:

```c
time: 12:00 -> RAPL    ->  2400 points
time: 12:02 -> RAPL    ->  2400 points
time: 12:04 -> RAPL    ->  2400 points
...
```

#### Example 2: RAPL + procfs (with different `flush_interval` values)

Now consider running two sources with different flush intervals:

```toml
[plugins.rapl]
# Interval between two RAPL measurements.
poll_interval = "5ms"
# Interval between two flushing of RAPL measurements.
flush_interval = "2s"
# Set to true to disable perf-events and always use the powercap sysfs.
no_perf_events = false

[plugins.procfs.processes.events]
# Interval between two measurements for event-driven process monitoring.
poll_interval = "500ms"
# How frequently should the processes information be flushed to the rest of the pipeline.
flush_interval = "4s"
# Which method to use to obtain memory statistics.
memory_mode = "quick"

# ...
```

RAPL continues to flush 2400 points every 2s as before. **In parallel**, procfs flushes (4s / 0.500s) × 9 = **72 points every 4s**.

> Each source flushes independently: the Alumet/OTEL plugin receives a separate `MeasurementBuffer` from each and issues a dedicated OTLP/gRPC request for each — concurrently. The resulting traffic looks like:

Reviewer: this is actually not totally true, as procfs will create one source per process, the number of batchs at the end depends on the number process watched by alumet (in a simple case with one process using alumet exec there's a single source but in usual watch mode you would have lot of them in //).

On its side, RAPL plugin always have only one source

Author: got it, I'll update the README.md
